netty/src/main/java/io/grpc/netty/NettyClientTransport.java

```diff
@@ -155,6 +165,13 @@ public Runnable start(Listener transportListener) {
     if (NioSocketChannel.class.isAssignableFrom(channelType)) {
       b.option(SO_KEEPALIVE, true);
     }
+
+    /**
+     * Proxy connections needs to happen first, unencrypted, before any other operation.
+     */
+    if (proxyHandler != null) {
+      b.handler(proxyHandler);
```

Comment:
There is only one handler in Bootstrap. The later call to `b.handler(negotiationHandler)` will overwrite this one.

The single handler is generally a `ChannelInitializer`. However, there are some multithreading issues we encountered while trying to use it (with how channels work and how we're using them, not the initializer itself; it was non-obvious and complex). So instead you need to extend `ProtocolNegotiator.AbstractBufferingHandler`.